File: sdk/py/fractary_core/docs/manager.py
"""
DocsManager - Framework-agnostic documentation management.

Provides documentation creation, search, and version management.
"""

from __future__ import annotations

import os
import re
from dataclasses import dataclass, field
from datetime import datetime
from pathlib import Path
from typing import Any, Literal, Optional

import yaml


DocFormat = Literal["markdown", "html", "pdf", "text"]


@dataclass
class DocMetadata:
    """Metadata for a document."""

    title: str
    author: Optional[str] = None
    version: str = "1.0.0"
    tags: list[str] = field(default_factory=list)
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


@dataclass
class Doc:
    """Represents a documentation file."""

    id: str
    content: str
    format: DocFormat
    metadata: DocMetadata
    path: str


class DocsManager:
    """Framework-agnostic documentation management.

    Provides creation, search, and management of documentation
    with support for multiple formats and versioning.
    """

    def __init__(self, config: Optional[dict[str, Any]] = None) -> None:
        """Initialize DocsManager with optional config.

        Args:
            config: Configuration dict with 'docs_dir' and 'default_format'
        """
        self.config = config or {}
        self.docs_dir = Path(self.config.get("docs_dir", ".fractary/docs"))
        self.docs_dir.mkdir(parents=True, exist_ok=True)
        self.default_format: DocFormat = self.config.get("default_format", "markdown")

    def _get_doc_path(self, doc_id: str, format: DocFormat) -> Path:
        """Get the file path for a document."""
        extensions = {
            "markdown": ".md",
            "html": ".html",
            "pdf": ".pdf",
            "text": ".txt",
        }
        return self.docs_dir / f"{doc_id}{extensions[format]}"

    def _get_metadata_path(self, doc_id: str) -> Path:
        """Get the metadata file path for a document."""
        return self.docs_dir / f"{doc_id}.meta.yaml"

    def create_doc(
        self,
        id: str,
        content: str,
        metadata: DocMetadata,
        format: Optional[DocFormat] = None,
    ) -> Doc:
        """Create a new documentation file.

        Args:
            id: Document identifier
            content: Document content
            metadata: Document metadata
            format: Document format (default: configured default)

        Returns:
            Created Doc object
        """
        doc_format = format or self.default_format
        doc_path = self._get_doc_path(id, doc_format)
        meta_path = self._get_metadata_path(id)

        # Update metadata timestamps
        full_metadata = DocMetadata(
            title=metadata.title,
            author=metadata.author,
            version=metadata.version,
            tags=metadata.tags,
            created_at=metadata.created_at or datetime.now(),
            updated_at=datetime.now(),
        )

        # Write document content
        doc_path.write_text(content, encoding="utf-8")

        # Write metadata
        meta_dict = {
            "title": full_metadata.title,
            "author": full_metadata.author,
            "version": full_metadata.version,
            "tags": full_metadata.tags,
            "created_at": full_metadata.created_at.isoformat() if full_metadata.created_at else None,
            "updated_at": full_metadata.updated_at.isoformat() if full_metadata.updated_at else None,
        }
        meta_path.write_text(yaml.dump(meta_dict), encoding="utf-8")

        return Doc(
            id=id,
            content=content,
            format=doc_format,
            metadata=full_metadata,
            path=str(doc_path),
        )

    def get_doc(self, doc_id: str) -> Optional[Doc]:
        """Get a document by ID.

        Args:
            doc_id: Document identifier

        Returns:
            Doc object or None if not found
        """
        # Try to find doc with any format
        for format in ["markdown", "html", "pdf", "text"]:
            doc_path = self._get_doc_path(doc_id, format)  # type: ignore
            if doc_path.exists():
                content = doc_path.read_text(encoding="utf-8")

                # Load metadata
                meta_path = self._get_metadata_path(doc_id)
                if meta_path.exists():
                    meta_dict = yaml.safe_load(meta_path.read_text(encoding="utf-8"))
                    metadata = DocMetadata(
                        title=meta_dict.get("title", doc_id),
                        author=meta_dict.get("author"),
                        version=meta_dict.get("version", "1.0.0"),
                        tags=meta_dict.get("tags", []),
                        created_at=(
                            datetime.fromisoformat(meta_dict["created_at"])
                            if meta_dict.get("created_at")
                            else None
                        ),
                        updated_at=(
                            datetime.fromisoformat(meta_dict["updated_at"])
                            if meta_dict.get("updated_at")
                            else None
                        ),
                    )
                else:
                    metadata = DocMetadata(title=doc_id)

                return Doc(
                    id=doc_id,
                    content=content,
                    format=format,  # type: ignore
                    metadata=metadata,
                    path=str(doc_path),
                )

        return None

    def update_doc(
        self,
        doc_id: str,
        content: Optional[str] = None,
        metadata: Optional[DocMetadata] = None,
    ) -> Optional[Doc]:
        """Update an existing document.

        Args:
            doc_id: Document identifier
            content: New content (optional)
            metadata: Updated metadata (optional)

        Returns:
            Updated Doc object or None if not found
        """
        doc = self.get_doc(doc_id)
        if not doc:
            return None

        # Update content if provided
        if content is not None:
            doc_path = self._get_doc_path(doc_id, doc.format)
            doc_path.write_text(content, encoding="utf-8")
            doc.content = content

        # Update metadata if provided
        if metadata:
            metadata.updated_at = datetime.now()
            meta_path = self._get_metadata_path(doc_id)
            meta_dict = {
                "title": metadata.title,
                "author": metadata.author,
                "version": metadata.version,
                "tags": metadata.tags,
                "created_at": metadata.created_at.isoformat() if metadata.created_at else None,
                "updated_at": metadata.updated_at.isoformat() if metadata.updated_at else None,
            }
            meta_path.write_text(yaml.dump(meta_dict), encoding="utf-8")
            doc.metadata = metadata

        return doc

    def delete_doc(self, doc_id: str) -> bool:
        """Delete a document.

        Args:
            doc_id: Document identifier

        Returns:
            True if deleted successfully
        """
        deleted = False

        # Delete all format variants
        for format in ["markdown", "html", "pdf", "text"]:
            doc_path = self._get_doc_path(doc_id, format)  # type: ignore
            if doc_path.exists():
                doc_path.unlink()
                deleted = True

        # Delete metadata
        meta_path = self._get_metadata_path(doc_id)
        if meta_path.exists():
            meta_path.unlink()
            deleted = True

        return deleted

    def list_docs(
        self,
        tag: Optional[str] = None,
        format: Optional[DocFormat] = None,
    ) -> list[Doc]:
        """List all documents with optional filtering.

        Args:
            tag: Filter by tag
            format: Filter by format

        Returns:
            List of Doc objects
        """
        docs = []
        seen_ids = set()

        for file_path in self.docs_dir.glob("*"):
            if file_path.suffix in [".md", ".html", ".pdf", ".txt"]:
                doc_id = file_path.stem
                if doc_id in seen_ids:
                    continue

                doc = self.get_doc(doc_id)
                if doc:
                    # Apply filters
                    if format and doc.format != format:
                        continue
                    if tag and tag not in doc.metadata.tags:
                        continue

                    docs.append(doc)
                    seen_ids.add(doc_id)

        return sorted(docs, key=lambda d: d.metadata.updated_at or datetime.min, reverse=True)

    def search_docs(self, query: str) -> list[Doc]:
        """Search documents by content or title.

        Args:
            query: Search query string

        Returns:
            List of matching Doc objects
        """
        results = []
        query_lower = query.lower()

        for doc in self.list_docs():
            # Search in title and content
            if (
                query_lower in doc.metadata.title.lower()
                or query_lower in doc.content.lower()
            ):
                results.append(doc)

        return results

File: sdk/js/src/docs/manager.ts
/**
 * @fractary/core - Docs Manager
 *
 * Documentation management and organization.
 */

import * as fs from 'fs';
import * as path from 'path';
import * as yaml from 'js-yaml';
import {
  Doc,
  DocFormat,
  DocMetadata,
  DocSearchQuery,
  DocsManagerConfig,
  MetadataMode,
} from './types';

/**
 * Frontmatter parsing result
 */
interface FrontmatterResult {
  metadata: DocMetadata;
  body: string;
}

/**
 * Documentation Manager - Create, manage, and search documentation
 */
export class DocsManager {
  private docsDir: string;
  private defaultFormat: DocFormat;
  private metadataMode: MetadataMode;

  constructor(config: DocsManagerConfig) {
    this.docsDir = config.docsDir;
    this.defaultFormat = config.defaultFormat || 'markdown';
    this.metadataMode = config.metadataMode || 'frontmatter';
    this.ensureDir(this.docsDir);
  }

  private ensureDir(dirPath: string): void {
    if (!fs.existsSync(dirPath)) {
      fs.mkdirSync(dirPath, { recursive: true });
    }
  }

  private getDocPath(id: string, format: DocFormat): string {
    const ext = this.getFileExtension(format);
    return path.join(this.docsDir, `${id}${ext}`);
  }

  private getMetadataPath(id: string): string {
    return path.join(this.docsDir, `${id}.meta.yaml`);
  }

  private getFileExtension(format: DocFormat): string {
    const extensions: Record<DocFormat, string> = {
      markdown: '.md',
      html: '.html',
      pdf: '.pdf',
      text: '.txt',
    };
    return extensions[format];
  }

  /**
   * Determine the effective metadata mode for a document format
   * Frontmatter only makes sense for markdown files
   */
  private getEffectiveMetadataMode(format: DocFormat): MetadataMode {
    if (format !== 'markdown') {
      return 'sidecar';
    }
    return this.metadataMode;
  }

  /**
   * Parse frontmatter from markdown content
   */
  private parseFrontmatter(content: string): FrontmatterResult {
    const frontmatterRegex = /^---\r?\n([\s\S]*?)\r?\n---\r?\n([\s\S]*)$/;
    const match = content.match(frontmatterRegex);

    if (!match) {
      // No frontmatter found, return content as-is with minimal metadata
      return {
        metadata: { title: '' },
        body: content,
      };
    }

    const frontmatterContent = match[1];
    const body = match[2];

    let metadata: DocMetadata;
    try {
      const parsed = yaml.load(frontmatterContent) as Record<string, unknown>;
      metadata = {
        title: (parsed.title as string) || '',
        description: parsed.description as string | undefined,
        authors: parsed.authors as string[] | undefined,
        tags: parsed.tags as string[] | undefined,
        docType: parsed.docType as string | undefined,
        status: parsed.status as string | undefined,
        version: parsed.version as string | undefined,
      };

      // Parse dates
      if (parsed.createdAt) {
        metadata.createdAt = new Date(parsed.createdAt as string);
      }
      if (parsed.updatedAt) {
        metadata.updatedAt = new Date(parsed.updatedAt as string);
      }

      // Copy any additional fields
      for (const [key, value] of Object.entries(parsed)) {
        if (!(key in metadata)) {
          metadata[key] = value;
        }
      }
    } catch {
      metadata = { title: '' };
    }

    return { metadata, body };
  }

  /**
   * Write frontmatter to markdown content
   */
  private writeFrontmatter(metadata: DocMetadata, body: string): string {
    // Prepare metadata object for YAML serialization
    const yamlMetadata: Record<string, unknown> = {};

    // Add standard fields
    if (metadata.title) yamlMetadata.title = metadata.title;
    if (metadata.description) yamlMetadata.description = metadata.description;
    if (metadata.authors && metadata.authors.length > 0) yamlMetadata.authors = metadata.authors;
    if (metadata.tags && metadata.tags.length > 0) yamlMetadata.tags = metadata.tags;
    if (metadata.docType) yamlMetadata.docType = metadata.docType;
    if (metadata.status) yamlMetadata.status = metadata.status;
    if (metadata.version) yamlMetadata.version = metadata.version;

    // Add dates
    if (metadata.createdAt) {
      yamlMetadata.createdAt = metadata.createdAt instanceof Date
        ? metadata.createdAt.toISOString()
        : metadata.createdAt;
    }
    if (metadata.updatedAt) {
      yamlMetadata.updatedAt = metadata.updatedAt instanceof Date
        ? metadata.updatedAt.toISOString()
        : metadata.updatedAt;
    }

    // Add any additional custom fields
    for (const [key, value] of Object.entries(metadata)) {
      if (
        value !== undefined &&
        !['title', 'description', 'authors', 'tags', 'docType', 'status', 'version', 'createdAt', 'updatedAt'].includes(key)
      ) {
        yamlMetadata[key] = value;
      }
    }

    const frontmatter = yaml.dump(yamlMetadata, { lineWidth: -1 });
    return `---\n${frontmatter}---\n\n${body}`;
  }

  /**
   * Extract body content from content that may have frontmatter
   */
  private extractBody(content: string): string {
    const result = this.parseFrontmatter(content);
    return result.body;
  }

  /**
   * Create a new document
   * @param id - Document identifier
   * @param content - Document body content (without frontmatter)
   * @param metadata - Document metadata
   * @param format - Document format (default: markdown)
   */
  async createDoc(
    id: string,
    content: string,
    metadata: DocMetadata,
    format?: DocFormat
  ): Promise<Doc> {
    const docFormat = format || this.defaultFormat;
    const docPath = this.getDocPath(id, docFormat);
    const effectiveMode = this.getEffectiveMetadataMode(docFormat);

    // Add timestamps
    const fullMetadata: DocMetadata = {
      ...metadata,
      createdAt: metadata.createdAt || new Date(),
      updatedAt: new Date(),
    };

    // Extract body content in case content already has frontmatter
    const bodyContent = this.extractBody(content);

    if (effectiveMode === 'frontmatter') {
      // Write content with embedded frontmatter
      const fullContent = this.writeFrontmatter(fullMetadata, bodyContent);
      fs.writeFileSync(docPath, fullContent, 'utf-8');
    } else {
      // Sidecar mode: separate content and metadata files
      const metaPath = this.getMetadataPath(id);
      fs.writeFileSync(docPath, bodyContent, 'utf-8');
      fs.writeFileSync(metaPath, yaml.dump(fullMetadata), 'utf-8');
    }

    return {
      id,
      content: bodyContent,
      format: docFormat,
      metadata: fullMetadata,
      path: docPath,
    };
  }

  /**
   * Get a document by ID
   * Supports both frontmatter and sidecar metadata modes
   */
  async getDoc(id: string): Promise<Doc | null> {
    // Try to find the doc with any format
    const formats: DocFormat[] = ['markdown', 'html', 'pdf', 'text'];

    for (const format of formats) {
      const docPath = this.getDocPath(id, format);
      if (fs.existsSync(docPath)) {
        const rawContent = fs.readFileSync(docPath, 'utf-8');
        const effectiveMode = this.getEffectiveMetadataMode(format);
        let metadata: DocMetadata = { title: id };
        let content: string;

        if (effectiveMode === 'frontmatter') {
          // Parse frontmatter from content
          const result = this.parseFrontmatter(rawContent);
          metadata = result.metadata.title ? result.metadata : { ...metadata, ...result.metadata };
          content = result.body;
        } else {
          // Sidecar mode: check for .meta.yaml file
          content = rawContent;
          const metaPath = this.getMetadataPath(id);
          if (fs.existsSync(metaPath)) {
            const metaContent = fs.readFileSync(metaPath, 'utf-8');
            metadata = yaml.load(metaContent) as DocMetadata;
          }
        }

        return {
          id,
          content,
          format,
          metadata,
          path: docPath,
        };
      }
    }

    return null;
  }

  /**
   * Update a document
   * @param id - Document identifier
   * @param content - New body content (without frontmatter)
   * @param metadata - Metadata fields to update (merged with existing)
   */
  async updateDoc(
    id: string,
    content: string,
    metadata?: Partial<DocMetadata>
  ): Promise<Doc | null> {
    const existingDoc = await this.getDoc(id);
    if (!existingDoc) {
      return null;
    }

    const updatedMetadata: DocMetadata = {
      ...existingDoc.metadata,
      ...metadata,
      updatedAt: new Date(),
    };

    const docPath = this.getDocPath(id, existingDoc.format);
    const effectiveMode = this.getEffectiveMetadataMode(existingDoc.format);

    // Extract body content in case content already has frontmatter
    const bodyContent = this.extractBody(content);

    if (effectiveMode === 'frontmatter') {
      // Write content with embedded frontmatter
      const fullContent = this.writeFrontmatter(updatedMetadata, bodyContent);
      fs.writeFileSync(docPath, fullContent, 'utf-8');
    } else {
      // Sidecar mode: separate content and metadata files
      const metaPath = this.getMetadataPath(id);
      fs.writeFileSync(docPath, bodyContent, 'utf-8');
      fs.writeFileSync(metaPath, yaml.dump(updatedMetadata), 'utf-8');
    }

    return {
      id,
      content: bodyContent,
      format: existingDoc.format,
      metadata: updatedMetadata,
      path: docPath,
    };
  }

  /**
   * Delete a document
   * Removes both the document file and any sidecar metadata file if it exists
   */
  async deleteDoc(id: string): Promise<boolean> {
    const doc = await this.getDoc(id);
    if (!doc) {
      return false;
    }

    const docPath = this.getDocPath(id, doc.format);
    const metaPath = this.getMetadataPath(id);

    if (fs.existsSync(docPath)) {
      fs.unlinkSync(docPath);
    }

    // Also delete sidecar metadata file if it exists (for backward compatibility)
    if (fs.existsSync(metaPath)) {
      fs.unlinkSync(metaPath);
    }

    return true;
  }

  /**
   * List all documents
   */
  async listDocs(): Promise<Doc[]> {
    if (!fs.existsSync(this.docsDir)) {
      return [];
    }

    const files = fs.readdirSync(this.docsDir);
    const docIds = new Set<string>();

    // Extract unique document IDs (remove extensions and .meta)
    for (const file of files) {
      if (file.endsWith('.meta.yaml')) {
        continue;
      }
      const id = file.replace(/\.(md|html|pdf|txt)$/, '');
      docIds.add(id);
    }

    const docs: Doc[] = [];
    for (const id of docIds) {
      const doc = await this.getDoc(id);
      if (doc) {
        docs.push(doc);
      }
    }

    return docs;
  }

  /**
   * Search documents
   */
  async searchDocs(query: DocSearchQuery): Promise<Doc[]> {
    const allDocs = await this.listDocs();
    let results = allDocs;

    // Filter by text
    if (query.text) {
      const searchText = query.text.toLowerCase();
      results = results.filter(
        (doc) =>
          doc.content.toLowerCase().includes(searchText) ||
          doc.metadata.title.toLowerCase().includes(searchText) ||
          doc.metadata.description?.toLowerCase().includes(searchText)
      );
    }

    // Filter by tags
    if (query.tags && query.tags.length > 0) {
      results = results.filter((doc) =>
        query.tags!.some((tag) => doc.metadata.tags?.includes(tag))
      );
    }

    // Filter by author
    if (query.author) {
      results = results.filter((doc) =>
        doc.metadata.authors?.includes(query.author!)
      );
    }

    // Filter by docType
    if (query.docType) {
      results = results.filter((doc) => doc.metadata.docType === query.docType);
    }

    // Filter by date range
    if (query.dateRange) {
      results = results.filter((doc) => {
        const docDate = doc.metadata.updatedAt || doc.metadata.createdAt;
        if (!docDate) return false;

        const date = new Date(docDate);
        if (query.dateRange!.from && date < query.dateRange!.from) {
          return false;
        }
        if (query.dateRange!.to && date > query.dateRange!.to) {
          return false;
        }
        return true;
      });
    }

    // Apply limit
    if (query.limit && query.limit > 0) {
      results = results.slice(0, query.limit);
    }

    return results;
  }

  /**
   * Check if a document exists
   */
  async docExists(id: string): Promise<boolean> {
    const doc = await this.getDoc(id);
    return doc !== null;
  }
}
